### PR TITLE
Bias Wave

### PR DESCRIPTION
**Pull request by msilvafe**
This is a draft PR for now...still lots to go but can start code level discussion here to move some of it off of slack.

The intention of this PR is to implement a new "bias wave" module that calculates equivalent parameters as the bias_step operation but using small sine waves on the bias line which can be processed using fourier analysis which is a bit cleaner in the presence of large halfwave plate synchronous signal.

### CODE REVIEW

**Comment by msilvafe** [comment]
Ok, I think this is ready to test. It currently can take data at multiple frequencies but only runs the analysis on the lowest frequency to calculate the DC parameters. We can add in analysis functions later for calculating time constants with multiple frequencies (there's some inline comments with placeholders for those things to go). 

This should probably come with an accompanying PR in socs to add an agent task but that can come after testing. I'm going to convert from draft to full PR so you all can start reviewing but I we will wait until we have taken some test data on SATP1 or SATP3 before merging.

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [inline sodetlib/operations/bias_wave.py]: I'm not sure what this is so... I probably did not use it before? perhaps you could clarify?

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [inline sodetlib/operations/bias_wave.py]: so one of the things worth checking is that when Yuhan and I wrote up our analysis notebooks we just used the bgmap from the most recent biasstep, usually was taken from before hwp spin-up. just wanted to confirm that is  what this code is able to do, or did you write it to generate a bgmap from the biaswave?

Comment (msilvafe) [reply sodetlib/operations/bias_wave.py]: See lines 109-114, if the argument `bg_map_file` provides a path to a bgmap then it will use that if not it will load the bgmap which is in the device config file which should be the most recent bgmap (which if you setup your night of observing this is likely also before you spun up the hwp for the night). There's no attempt to calculate a bgmap from the bias wave measurements here.

File: sodetlib/operations/bias_wave.py
Comment (msilvafe) [reply sodetlib/operations/bias_wave.py]: Oh ya...sorry I should probably just delete this out. I think it's sort of a vestige from an old version of the bias_step analysis where we would rely on this parameter to calculate the parameters differently depending if the bias step was taken when in the transition or not. Now this is estimated from the data directly and analysis updated accordingly.

Comment (bkelle) [reply sodetlib/operations/bias_wave.py]: gotcha gotcha, thanks for clarifying!

**Comment by bkelle** [comment]
Spent some time looking through this this morning. I think it looks good and I agree its ready to test as soon as we have a telescope we can try to run it on. That being said i'm not a total expert when it comes to sodetlib code writing. But I think this captures what Yuhan and I were doing in notebooks. Do I need to approve this commit before being able to test it, or should we test it and then approve after that?

**Comment by msilvafe** [comment]
> Spent some time looking through this this morning. I think it looks good and I agree its ready to test as soon as we have a telescope we can try to run it on. That being said i'm not a total expert when it comes to sodetlib code writing. But I think this captures what Yuhan and I were doing in notebooks. Do I need to approve this commit before being able to test it, or should we test it and then approve after that?

Please test on this branch, I'm not confident that the data taking won't just fail because I have a missed colon at this point...(I mostly don't think that's true bc I did run a linter on this but I think you get my point). So checkout this branch on the server you want to test on then run the data taking and analysis fn's and confirm that everything looks good. Post whatever plots/etc are useful from your testing here in the review to show that you've tested that it works and that can serve are your approval to merge.

**Comment by bkelle** [comment]
Sounds good Max. We will do that. Hopefully we can work it in to the SATp3 schedule when observing again, I think things have been down for a few days? In any case I'll interface with Yuhan, Sam, and folks at the site to test in this branch. thanks!

**Comment by msilvafe** [comment]
> Sounds good Max. We will do that. Hopefully we can work it in to the SATp3 schedule when observing again, I think things have been down for a few days? In any case I'll interface with Yuhan, Sam, and folks at the site to test in this branch. thanks!

Ok, sounds good! Let me know when you get a chance to take a stab at it. I can try to be available to help debug any issues that arise.

**Review by jlashner — CHANGES_REQUESTED** [review]
Thanks for this! I'm really glad it works, and though there is still work to be done testing it and making sure it is correct, I'm not opposed to merging this with a few docs requests:

- Can we write a nice docstring for the BiasWaveAnalysis class? In particular, something that documents the saved results, their shapes, etc.
- Can we try to remove as much vistigial "bias-step" code and docs as possible? There are a few things in here that are copied over from the bias step module (both in code and in docs), which make it difficult to know what is actually important for this analysis vs what is just copied over. I think if we can do a once over and try to remove as much of that as possible that would be great

Also, I question whether the analysis, as written, will function correctly, but fixing that can wait til a proper study is done and put into a second PR.

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:94]: Hi Jack, I'll work on this! Yeah, was planning to fill this out before merging.

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:261]: I think this is basically the line that caused all the problems we later had to fix yesterday @msilvafe . Do we want to do something else here, or leave it as is knowing that we modified things elsewhere to accommodate this?

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:525]: running each frequency on each BL for 3 seconds

File: sodetlib/operations/bias_wave.py
Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:424]: That's true...we need a way to check the phase of dItes relative to dIbias (would be easier with directly projecting out the frequency instead of taking from the periodogram)

File: sodetlib/operations/bias_wave.py
Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:525]: Depends on duration, and number of freqs arguments. Maybe @bkelle can give you the answer on the 2-frequencies and 3 sec duration test run we did and could probably do some scaling tests (for both time to run and data size) if we can get more time on satp3.

File: sodetlib/operations/bias_wave.py
Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:261]: Ya I think probably we want to change this to ints and then get rid of all of the hacky things we did later on to convert the values from this array to ints (and fix any checks we have for nans to not be based off of this array).

Comment (bkelle) [reply sodetlib/operations/bias_wave.py:261]: There's no nan equivalent int- would it be better to assign 0's (or maybe some arbitrarily high number) and then have an if statement so it doesn't analyze those values?

Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:261]: well `start_times` is a float. How about you just use that everywhere instead of `start_idxs`?

Comment (bkelle) [reply sodetlib/operations/bias_wave.py:261]: actually I can just put an if statement in so it fills amplitudes for empty slice BL. is it better to fill with 0 or np.nan amplitudes?

Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:261]: I'm not sure what you mean @bkelle/are concerned by I think amplitudes are already filled with nans right? (I think nan are preferable).

Comment (bkelle) [reply sodetlib/operations/bias_wave.py:261]: Oh I was saying that idk how the get_amplitudes works on a zero-length slice. so I can write an "if statement" to only calculate amplitudes for non-zero length slices,  and else assign amplitudes with np.nan

Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:261]: Is there a way that you think that could happen?

Comment (bkelle) [reply sodetlib/operations/bias_wave.py:261]: actually I think this is fine, testing the periodogram func it doesnt break if you feed it an empty array it just spits out nothing. so I'm not gonna change anything for now beyond just changing the .astype(int) fixes.

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:328]: fixing this in upcoming commit

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:132]: yup seconding max's comment here!

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:103]: deleting this __init__ section in upcoming commit

Comment (msilvafe) [reply sodetlib/operations/bias_wave.py]: I think mostly just removing `self.transition_range` I feel like we still want the rest.

Comment (bkelle) [reply sodetlib/operations/bias_wave.py]: got it and fixed in most recent commit!

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:417]: I vaguely recall Yuhan and I thinking about this a while back... I'll talk with him once he has a bit of time and I can refresh on what we discussed here/potentially change this

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:518]: It looks like yesterday for 2 frequencies it took about 2 mins! Which is basically consistent with bias steps then, ~1 min per frequency. I guess that means we should think about the data collection paradigm for i.e. time constant analysis with multiple frequencies as this could get large with like 7 frequencies or something like that.

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:430]: I just committed some changes meant to fix this with the help of Max. Let me know what y'all think.

Comment (msilvafe) [reply sodetlib/operations/bias_wave.py:430]: Did you test it and see that you got out negative values when you expected?

Comment (jlashner) [reply sodetlib/operations/bias_wave.py:430]: Maybe I'm misunderstanding... but I don't think the changes would work. Phase info isn't included in dItes or dIbias (since they're computed via sqrt(Pxx)), so I don't think its possible to extract the phase like this.

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:295]: oh yeah good catch thanks

File: sodetlib/operations/bias_wave.py
Comment (bkelle) [reply sodetlib/operations/bias_wave.py:470]: Have corrected calculation of amplitudes to allow for distinction between in- and out- of transition in recent commits.

**Review by jlashner — APPROVED** [review]
Thanks for the doc updates, I'm fine with merging this for now. 

I'm still not 100% confident in the DC param estimation and phase determination. In particular, like max said, there are better ways to figure out whether the signals are in phase that we should think about... 

Before using in operation, I'd like to see a study comparing directly DC params computed with bias waves with those from bias steps in and out of transition. Does that exist right now? If so can we link the presentation / confluence page? Thanks!

**Pull request merged**
